### PR TITLE
Create new threadpool when operating from thread

### PR DESCRIPTION
**Pull request by mrocklin**
This cleans up some issues in the threaded scheduler when operating from other threads.

1.  https://github.com/dask/distributed/pull/441 shows that dask.threaded.get can block when being called from other threads
2.  We were creating but not cleaning up threadpools when using `num_workers=` keyword argument

Now we keep a pool of pools around, keyed by thread and num_workers.  These get cleaned up at every call to get.

Reported by @AlbertDeFusco

### CODE REVIEW

**Comment by mrocklin** [comment]
I have a test for this in the distributed repository.  I'll create something here soon as well.

**Comment by mrocklin** [comment]
@jcrist @shoyer this change to the threaded scheduler could use your review if you have the time

File: dask/threaded.py [hunk 1]
```diff
@@ -45,16 +52,28 @@ def get(dsk, result, cache=None, num_workers=None, **kwargs):
...
-        if num_workers:
-            pool = ThreadPool(num_workers)
-        else:
+        if num_workers is None and thread is main_thread:
             pool = default_pool
+        elif thread in pools and num_workers in pools[thread]:
+            pool = pools[thread][num_workers]
+        else:
+            pool = ThreadPool(num_workers)
+            pools[thread][num_workers] = pool
```
Comment (shoyer) [inline dask/threaded.py]: Don't you need a lock here? I'm imagining a race condition where multiple threads try to create thread pools at the same time, e.g., there wasn't an existing pool when the check on line 61 was run, but there is when you overwrite the dictionary key.

File: dask/threaded.py [hunk 2]
```diff
@@ -5,21 +5,28 @@
...
-
 def _thread_get_id():
     return current_thread().ident
 
 
+default_pool = ThreadPool()
+default_thread = _thread_get_id()
+main_thread = current_thread()
+
+pools = defaultdict(dict)
```
Comment (shoyer) [inline dask/threaded.py]: I would probably just use a dict keyed on `(thread_id, num_workers)` tuples

Comment (mrocklin) [inline dask/threaded.py]: During cleanup it is convenient to have all threadpools indexed by thread.  Could flatten though.  In the common case this isn't large.

File: dask/threaded.py (hunk 1 above)
Comment (mrocklin) [inline dask/threaded.py]: I've added a lock fairly conservatively around the all `pools` management code

**Comment by mrocklin** [comment]
Merging soon if no comments

**Comment by shoyer** [comment]
This seems sane to me now.

**Pull request merged**